### DISCUSSION TITLE
Dynamic API Reference

### DISCUSSION THREAD

**Discussion by benhaynes** [directus/directus#3085] [Feature Requests]
I propose adding a module or Settings section that uses our API Reference structure (in markdown or OAS) to generate a tailored set of API Docs based on the project's schema. So each Collection and Field is looped over to have dynamic endpoint information.

**Comment by Nitwel** [comment]
Love that idea. ~~Should be doable and straigt forward to implement in theory.
The question is if we want to combine that with GH actions so they get updated automaticly?~~

**Reply by rijkvanzanten** [reply]
> Would be very easy and straigt forward to implement.

Doubt it 😉 

What would your step by step plan be to implement this @Nitwel?

**Reply by benhaynes** [reply]
Is it possible to render this on the fly instead of generating it? If not, perhaps we just have a re-generate button?

**Reply by rijkvanzanten** [reply]
Yeah, I'd say we'd want to import the latest API spec file (openapi) on build time, and use that to generate out the dynamic API view in runtime

**Reply by Nitwel** [reply]
~~Yep, generating this on the fly is very costly to do.
My step by step plan would be:~~
- ~~use the swagger parser to get all information needed~~
- ~~format the data into a md format~~
- ~~import / include these generated files into docs~~

~~It mostly is just reading the data you want and then printing it to the md file.~~

**Reply by rijkvanzanten** [reply]
@Nitwel Are we talking about the same thing? It sounds like you're talking about the API reference in the docs. We're talking about a section in settings that shows your endpoints on the fly with a way to test the IO

**Reply by Nitwel** [reply]
Yea, I mixed these things together. Now I get why you questioned my statements...

**Reply by benhaynes** [reply]
This is how I'm imagining the Docs Module being structured. Thoughts @Nitwel @rijkvanzanten ? Am I missing any API Reference sections?
```
/docs
    /getting-started
    /explanation
    /concepts
    /guides
    /api-reference
        /introduction
        /auth
        /errors
        /parameters
        /relational data
        ---
        /all dynamic endpoints
        ---
        /all static endpoints
```

**Reply by Nitwel** [reply]
We could also leave out objects and display them directly inside the static ones. Same for parameters and errors.

But lets just do everything as a link, see how it performs compared to the old way of doing it and when the old way was better, we change that?

**Reply by benhaynes** [reply]
I'm fine if objects only go within the endpoints... but parameters have a LOT of options/explanation, so I think it's best to reference them in the appropriate endpoints, but link to a dedicated page explaining each in more detail.

**Reply by rijkvanzanten** [reply]
@benhaynes Having a dedicated `/parameters` page under `/api-reference` might not make that much sense 🤔 Unless we want to list every endpoint that a param can be used at next to the param itself

**Reply by Nitwel** [reply]
Lets keep objects/schemas, parameters, responses on a single endpoint page and use menu/modal to show more details!

**Comment by benhaynes** [comment]
Thinking of adding SVGs inline into the Docs... this is a test to see if that works:

<svg width="35" height="40" fill="none" xmlns="http://www.w3.org/2000/svg">
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M16.86 18.92c.39-.22.86-.22 1.25 0l16.23 9.38a1.25 1.25 0 010 2.16l-16.23 9.37a1.25 1.25 0 01-1.88-1.08V20.01c0-.45.24-.86.63-1.09zm1.87 3.25v14.42l12.49-7.21-12.49-7.21z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M16.86.18c.39-.22.86-.22 1.25 0l16.23 9.37a1.25 1.25 0 010 2.17l-16.23 9.37A1.25 1.25 0 0116.23 20V1.26c0-.45.24-.86.63-1.08zm1.87 3.25v14.41l12.49-7.2-12.49-7.21z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M.62 9.55c.4-.22.87-.22 1.25 0l16.24 9.37a1.25 1.25 0 010 2.17L1.87 30.46A1.25 1.25 0 010 29.38V10.63c0-.44.24-.86.62-1.08zM2.5 12.8V27.2l12.48-7.2L2.5 12.8z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M34.34 9.55c.39.22.63.64.63 1.08v18.75c0 .44-.24.86-.63 1.08l-16.23 9.37a1.25 1.25 0 01-1.88-1.08V20c0-.44.24-.85.63-1.08l16.23-9.37c.39-.22.87-.22 1.25 0zm-15.6 11.18v15.86l13.73-7.93V12.8l-13.74 7.93z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M.62 9.55c.4-.22.87-.22 1.25 0l16.24 9.37c.38.23.62.64.62 1.08v18.75a1.25 1.25 0 01-1.87 1.08L.62 30.46A1.25 1.25 0 010 29.38V10.63c0-.44.24-.86.62-1.08zM2.5 12.8v15.86l13.73 7.93V20.73L2.5 12.8z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M16.86.17c.4-.23.87-.23 1.25 0l16.24 9.37a1.25 1.25 0 010 2.16L18.1 21.08c-.38.22-.86.22-1.25 0L.63 11.7a1.25 1.25 0 010-2.16L16.86.17zM3.76 10.62l13.73 7.93 13.73-7.93L17.5 2.7 3.76 10.62z" fill="#000"/>
  <path fill-rule="evenodd" clip-rule="evenodd" d="M.63 9.55c.39-.22.86-.22 1.25 0l15.62 9.01 15.62-9A1.25 1.25 0 0135 10.62v18.75a1.25 1.25 0 11-2.5 0V12.8l-14.37 8.29c-.4.22-.87.22-1.25 0L2.5 12.79v16.59a1.25 1.25 0 11-2.5 0V10.63c0-.44.23-.86.62-1.08z" fill="#000"/>
</svg>

**Reply by rijkvanzanten** [reply]
That's a no 🤓

**Reply by Nitwel** [reply]
How does it look?

**Reply by rijkvanzanten** [reply]
Invisible; GH strips out inline SVGs in it's markdown rendering, this was the test:

```
Thinking of adding SVGs inline into the Docs... this is a test to see if that works:

<svg width="35" height="40" fill="none" xmlns="http://www.w3.org/2000/svg">
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M16.86 18.92c.39-.22.86-.22 1.25 0l16.23 9.38a1.25 1.25 0 010 2.16l-16.23 9.37a1.25 1.25 0 01-1.88-1.08V20.01c0-.45.24-.86.63-1.09zm1.87 3.25v14.42l12.49-7.21-12.49-7.21z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M16.86.18c.39-.22.86-.22 1.25 0l16.23 9.37a1.25 1.25 0 010 2.17l-16.23 9.37A1.25 1.25 0 0116.23 20V1.26c0-.45.24-.86.63-1.08zm1.87 3.25v14.41l12.49-7.2-12.49-7.21z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M.62 9.55c.4-.22.87-.22 1.25 0l16.24 9.37a1.25 1.25 0 010 2.17L1.87 30.46A1.25 1.25 0 010 29.38V10.63c0-.44.24-.86.62-1.08zM2.5 12.8V27.2l12.48-7.2L2.5 12.8z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M34.34 9.55c.39.22.63.64.63 1.08v18.75c0 .44-.24.86-.63 1.08l-16.23 9.37a1.25 1.25 0 01-1.88-1.08V20c0-.44.24-.85.63-1.08l16.23-9.37c.39-.22.87-.22 1.25 0zm-15.6 11.18v15.86l13.73-7.93V12.8l-13.74 7.93z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M.62 9.55c.4-.22.87-.22 1.25 0l16.24 9.37c.38.23.62.64.62 1.08v18.75a1.25 1.25 0 01-1.87 1.08L.62 30.46A1.25 1.25 0 010 29.38V10.63c0-.44.24-.86.62-1.08zM2.5 12.8v15.86l13.73 7.93V20.73L2.5 12.8z" fill="#000"/>
  <path opacity=".25" fill-rule="evenodd" clip-rule="evenodd" d="M16.86.17c.4-.23.87-.23 1.25 0l16.24 9.37a1.25 1.25 0 010 2.16L18.1 21.08c-.38.22-.86.22-1.25 0L.63 11.7a1.25 1.25 0 010-2.16L16.86.17zM3.76 10.62l13.73 7.93 13.73-7.93L17.5 2.7 3.76 10.62z" fill="#000"/>
  <path fill-rule="evenodd" clip-rule="evenodd" d="M.63 9.55c.39-.22.86-.22 1.25 0l15.62 9.01 15.62-9A1.25 1.25 0 0135 10.62v18.75a1.25 1.25 0 11-2.5 0V12.8l-14.37 8.29c-.4.22-.87.22-1.25 0L2.5 12.79v16.59a1.25 1.25 0 11-2.5 0V10.63c0-.44.23-.86.62-1.08z" fill="#000"/>
</svg>
```

**Comment by licitdev** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
